ta/pkcs11/src/processing_rsa.c

```diff
@@ -10,8 +10,110 @@
 #include <tee_internal_api_extensions.h>
 
 #include "attributes.h"
+#include "object.h"
 #include "processing.h"
 
+enum pkcs11_rc load_tee_rsa_key_attrs(TEE_Attribute **tee_attrs,
+				      size_t *tee_count,
+				      struct pkcs11_object *obj)
+{
+	TEE_Attribute *attrs = NULL;
+	size_t count = 0;
+	enum pkcs11_rc rc = PKCS11_CKR_GENERAL_ERROR;
+	void *a_ptr = NULL;
+
+	assert(get_key_type(obj->attributes) == PKCS11_CKK_RSA);
+
+	switch (get_class(obj->attributes)) {
+	case PKCS11_CKO_PUBLIC_KEY:
+		attrs = TEE_Malloc(3 * sizeof(TEE_Attribute),
+				   TEE_USER_MEM_HINT_NO_FILL_ZERO);
+		if (!attrs)
+			return PKCS11_CKR_DEVICE_MEMORY;
+
+		if (pkcs2tee_load_attr(&attrs[count], TEE_ATTR_RSA_MODULUS,
+				       obj, PKCS11_CKA_MODULUS))
+			count++;
+
+		if (pkcs2tee_load_attr(&attrs[count],
+				       TEE_ATTR_RSA_PUBLIC_EXPONENT, obj,
+				       PKCS11_CKA_PUBLIC_EXPONENT))
+			count++;
+
+		if (count == 2)
+			rc = PKCS11_CKR_OK;
+
+		break;
+
+	case PKCS11_CKO_PRIVATE_KEY:
+		attrs = TEE_Malloc(8 * sizeof(TEE_Attribute),
+				   TEE_USER_MEM_HINT_NO_FILL_ZERO);
+		if (!attrs)
+			return PKCS11_CKR_DEVICE_MEMORY;
+
+		if (pkcs2tee_load_attr(&attrs[count], TEE_ATTR_RSA_MODULUS,
+				       obj, PKCS11_CKA_MODULUS))
+			count++;
+
+		if (pkcs2tee_load_attr(&attrs[count],
+				       TEE_ATTR_RSA_PUBLIC_EXPONENT, obj,
+				       PKCS11_CKA_PUBLIC_EXPONENT))
+			count++;
+
+		if (pkcs2tee_load_attr(&attrs[count],
+				       TEE_ATTR_RSA_PRIVATE_EXPONENT, obj,
+				       PKCS11_CKA_PRIVATE_EXPONENT))
+			count++;
+
+		if (count != 3)
+			break;
+
+		if (get_attribute(obj->attributes, PKCS11_CKA_PRIME_1, NULL,
```

Comment:
here `a_ptr` is always null. I think this is a typo and you expected
```c
		if (get_attribute(obj->attributes, PKCS11_CKA_PRIME_1, &a_ptr,
				  NULL) || !a_ptr) {
			rc = PKCS11_CKR_OK;
			break;
```

Note that since `get_attribute()` could return something else  than OK or NOT_FOUND, is would be better to be strict on its return value:
```c
		rc = get_attribute_ptr(obj->attributes, PKCS11_CKA_PRIME_1,
				       &a_ptr, NULL);
		switch (rc) {
		case PKCS11_CKR_OK:
			if (!a_ptr)
				goto out;
			break;
		case PKCS11_RV_NOT_FOUND:
			rc = PKCS11_CKR_OK;
			goto out;
		default:
			goto out;

		}


		if (pkcs2tee_load_attr(&attrs[count], TEE_ATTR_RSA_PRIME1, obj,
				       PKCS11_CKA_PRIME_1))
			count++;
		...

Reply:
OK

Reply:
looks like the above is buggy... ans switch/case not very convenient.
maybe simply:
```c
		rc = get_attribute_ptr(obj->attributes, PKCS11_CKA_PRIME_1,
				       &a_ptr, NULL);
		if (rc != PKCS11_CKR_OK && rc != PKCS11_RV_NOT_FOUND)
			break;
		if (rc == PKCS11_RV_NOT_FOUND || !a_ptr) {
			rc = PKCS11_CKR_OK;
			break;
		}

  		...

Reply:
Can use that construct.